### PR TITLE
feat(models): Baichuan

### PR DESCRIPTION
**Pull request by hetaoBackend**
The baichuan-7B is a widely-spread open-source model, I'd like to integrate it with OpenLLM.

Baichuan Introduction
https://github.com/baichuan-inc/Baichuan-7B/blob/main/README_EN.md

HuggingFace Link
https://huggingface.co/baichuan-inc/Baichuan-7B

### CODE REVIEW

**Comment by aarnphm** [comment]
Hi there. Thanks for adding this. Kindly ask you to add a test case under tests/models_test.py as test_baichua_implementation

**Comment by hetaoBackend** [comment]
> Hi there. Thanks for adding this. Kindly ask you to add a test case under tests/models_test.py as test_baichua_implementation

Cool! I will add the test for the new model.

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (aarnphm) [inline src/openllm/models/baichuan/modeling_baichuan.py:42]: U dont need to do this, since you already set trust_remote_code under BaichuanConfig. Unless there are special way to load this tokenizer, load_tokenizer js not needed here.

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (aarnphm) [inline src/openllm/models/baichuan/modeling_baichuan.py:48]: From the quick look of it, it seems like some of the default models does support different prompt template. Can we so something similar to mpt prompt parsing here?

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (aarnphm) [inline src/openllm/models/baichuan/modeling_baichuan.py:70]: Will probably need to also use torch.amp.autocast here since baichuan does support lower precision inference.

**Review by aarnphm — CHANGES_REQUESTED** [review]
Can u also add the import under __init__.py for Baichuan and BaichuanConfig? So that we can do the following:

```python
import openllm

Config = openllm.BaichuanConfig.from_pretrained()

Llm = openllm.Baichuan.from_pretrained()
```

You can take a look at the __init__.py for other example.

Otherwise everything else LGTM 🙂

**Comment by aarnphm** [comment]
I will take another look once im back home 🙂

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (hetaoBackend) [reply src/openllm/models/baichuan/modeling_baichuan.py:70]: Cool! All the things make sense to me. I will revise the code based on the comments! 👍

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (hetaoBackend) [reply src/openllm/models/baichuan/modeling_baichuan.py:42]: Gotcha!

Comment (hetaoBackend) [reply src/openllm/models/baichuan/modeling_baichuan.py]: I've removed these two load functions. Thanks!

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (hetaoBackend) [reply src/openllm/models/baichuan/modeling_baichuan.py]: I've enabled `torch.amp.autocast` here. Please kindly review it again. Thanks!

File: src/openllm/models/baichuan/modeling_baichuan.py
Comment (hetaoBackend) [reply src/openllm/models/baichuan/modeling_baichuan.py:42]: Since the prompt template is kind of simple right now, I implemented a way like `falcon` and `flan_t5` do. Does it make sense to u?

Comment (aarnphm) [reply src/openllm/models/baichuan/modeling_baichuan.py:42]: Ye I think the prompt can just be very simple for now, that's fine.

**Review by aarnphm — APPROVED** [review]
LGTM. Thanks again for the contribution. I will wait for the tests to run and will merge this after the tests finish.

**Comment by hetaoBackend** [comment]
> LGTM. Thanks again for the contribution. I will wait for the tests to run and will merge this after the tests finish.

Thanks a lot! @aarnphm. However, I noticed that the tests failed, anything else I could help with?

**Comment by aarnphm** [comment]
@hetaoBackend no worries I got it covered 👍

File: src/openllm/models/baichuan/configuration_baichuan.py
Comment (hetaoBackend) [inline src/openllm/models/baichuan/configuration_baichuan.py:39]: Hey @aarnphm, it is cool to add baichuan-13b models. However, I noticed that `cpm-kernels` is required by the `baichuan-13b` models, do we need to add `cpm-kernels` as the requirements for baichuan model? Any thoughts?

Reference: https://github.com/baichuan-inc/Baichuan-13B/blob/main/requirements.txt#L3

Comment (aarnphm) [reply src/openllm/models/baichuan/configuration_baichuan.py:39]: oh i see, let me fix that. I can add it to requirements

Comment (hetaoBackend) [reply src/openllm/models/baichuan/configuration_baichuan.py:39]: I've created a PR https://github.com/bentoml/OpenLLM/pull/117 to fix the dependency issue, it would be genuinely appreciate if u could kindly review it.

**Pull request merged**
